Run-benchmark should be able to checkout out a fixed version of benchmark from Git/GitHub

dewei-zhu opened the pull request:
#### 07fadde3a7d01ced085dc1a968d48b25e4e77b5c
<pre>
&apos;run-benchmark&apos; script should not fail when script is out of a git checkout.
<a href="https://bugs.webkit.org/show_bug.cgi?id=245452">https://bugs.webkit.org/show_bug.cgi?id=245452</a>
rdar://99465558

Reviewed by NOBODY (OOPS!).

Fix a bug that &apos;run-benchmark&apos; raises an exception when script is not inside a git checkout.

* Tools/Scripts/webkitpy/benchmark_runner/benchmark_builder.py:
(BenchmarkBuilder._local_git_archive_eligible):
</pre>

shallawa commented on Tools/Scripts/webkitpy/benchmark_runner/data/plans/motionmark1.2.1.plan:
```diff
@@ -0,0 +1,157 @@
+    "github_subtree": {
+        "url": "https://api.github.com/repos/WebKit/WebKit/git/trees/23431f74c4e13875561a7778e9d98628d4293ea0",
+        "truncated": false,
+        "tree": [
```
Do we have to list all the files in the plan? Can't Github API help us enumerating all files under a a certain directory? Otherwise we have to remember to update this tree if we add/remove a file.

dewei-zhu replied:
We can, but there is 60 requests/hour limit on `api.github.com`. We are trying to avoid hitting it by caching the response here since the response for a given committed commit should never change.
The plan file can work without `github_subtree`, but engineer will be effectively subject to "no more than 60 times per hour" limit.
`github_subtree` field can be generated using `plan_file_optimizer.py`

shallawa commented on Tools/Scripts/webkitpy/benchmark_runner/github_downloader.py:
```diff
@@ -0,0 +1,195 @@
+
+    @classmethod
+    def is_directory(cls, entry):
+        return entry['mode'].startswith('040')
```
why can't we use return entry['type'] == 'tree'?

dewei-zhu replied:
I guess we can, using `mode` is equivalent according to git documentation.

dewei-zhu replied:
Updated using `entry['type'] == 'tree'`.

sundiamonde commented on Tools/Scripts/webkitpy/benchmark_runner/data/plans/jetstream2.1.plan:
```diff
@@ -1,7 +1,520 @@
+        "url": "https://api.github.com/repos/WebKit/WebKit/git/trees/daa4085ae24f00a4bbefaa2f1abf02ab31b57afd",
+        "truncated": false,
+        "tree": [
+            {"path": "ARES-6", "mode": "040000", "type": "tree"},
```
Is there really no way to just specify directories? Probably won't matter for the benchmarks but we'll miss stuff in other uses

dewei-zhu replied:
This field should be generated using `plan_file_optimizer.py`.

sundiamonde commented on Tools/Scripts/webkitpy/benchmark_runner/github_downloader.py:
```diff
@@ -0,0 +1,195 @@
+        _log.info(f'Checking out content under {base_relpath_in_repo}:')
+        result_by_path = {}
+
+        with Pool(10) as pool:
```
hmm our Python2 implementation purposely had a broken implementation of Pool.  Thats why we don't use it anywhere.  Not sure if anything changed for Python3 so using this is risky

dewei-zhu replied:
Per discussion, python3 uses `spawn` on macOS by default since python3.8.
https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods

sundiamonde commented on Tools/Scripts/webkitpy/benchmark_runner/plan_file_optimizer.py:
```diff
@@ -0,0 +1,46 @@
+
+    def drop_keys(dictionary, keys_to_drop):
+        for key in keys_to_drop:
+            dictionary.pop(key, None)
```
nit: add a blank line here

dewei-zhu replied:
Will do.

sundiamonde commented on Tools/Scripts/webkitpy/benchmark_runner/benchmark_builder.py:
```diff
@@ -95,6 +103,32 @@ def _checkout_with_subversion(self, subversion_url):
+            return False
+        if not self.LOCAL_GIT_ARCHIVE_SCHEMA.match(self._plan['local_git_archive']):
+            return False
+        is_git_checkout = subprocess.check_output(['git', '-C', os.path.dirname(__file__), 'rev-parse',
```
will this fail if the path doesn't exist?

dewei-zhu replied:
No, because the path doesn't exist in current commit does not necessarily mean it doesn't exist in specified commit.
Since the plan file is manually created by us, if anything goes wrong, we should be able to notice and fix the issue very soon.

dewei-zhu replied:
Per discussion `os.path.dirname(__file__)` will always work unless we archive entire python into a zip executable.

dewei-zhu committed:
Run-benchmark should be able to checkout out a fixed version of benchmark from Git/GitHub
https://bugs.webkit.org/show_bug.cgi?id=244704
rdar://99465558

Reviewed by Stephanie Lewis.

Add 'local_git_archive' and 'github_source' schema to plan file so that we can download
a subdirectory from a Git repository or from GitHub.
Add 'github_subtree' schema to plan file as a cached and compressed response from GitHub
tree API. This can avoid hitting rate limit in GitHub API.
Write a module to support downloading subdirectories from GitHub.
Fix JetStream2.1 plan file to checkout the right version of JetStream.
Added MotionMark1.2.1 plan file.

* Tools/Scripts/webkitpy/benchmark_runner/benchmark_builder.py: Added support for 'local_git_archive'
and 'github_source'.
(BenchmarkBuilder):
(BenchmarkBuilder.__enter__):
(BenchmarkBuilder._download_from_github):
(BenchmarkBuilder._local_git_archive_eligible):
(BenchmarkBuilder._prepare_content_from_local_git_archive):
* Tools/Scripts/webkitpy/benchmark_runner/data/plans/jetstream2.1.plan:
* Tools/Scripts/webkitpy/benchmark_runner/data/plans/motionmark1.2.1.plan: Added.
* Tools/Scripts/webkitpy/benchmark_runner/github_downloader.py: Added this module for downloading a
subdirectory from GitHub using GitHub API and GitHub raw content endpoint.
(GithubDownloadTask):
(GithubDownloadTask.__init__):
(GithubDownloadTask.execute):
(GithubDownloadTask.subtree):
(retry_on_exception):
(retry_on_exception.decorator):
(retry_on_exception.decorator.wrapped):
(GithubDownloader):
(GithubDownloader.__init__):
(GithubDownloader.is_directory):
(GithubDownloader.is_file):
(GithubDownloader.is_symlink):
(GithubDownloader.session):
(GithubDownloader.download_content):
(GithubDownloader.subtree_for_path_with_commit):
(GithubDownloader._download_subtree_multiprocessing):
(GithubDownloader.walk):
(GithubDownloader._walk):
(GithubDownloader._subtree):
(GithubDownloader._commit):
(GithubDownloader._raw_content_url):
(GithubDownloader._download_raw_content):
(GithubDownloader._resolve_symlink):
(GithubDownloader._request):
* Tools/Scripts/webkitpy/benchmark_runner/plan_file_optimizer.py: Added this script
to cache the response from GitHub tree API for a given commit of a benchmark. This
will avoid hit any GitHub API, thus avoid hitting GitHub API rate limit.
(optimize_plan_file):
(optimize_plan_file.drop_keys):
(main):

Canonical link: https://commits.webkit.org/254124@main
